add packaging guide (#12)
swift-server/guides: packaging.md

# Packaging Applications for Deployment

> Once the application is built for production, it still needs to be packaged before it can be deployed to servers. There are several strategies for packaging Swift applications for deployment.

Reviewer: ```suggestion
Once an application is built for production, it still needs to be packaged before it can be deployed to servers. There are several strategies for packaging Swift applications for deployment.
```

## Docker

One of the most popular ways to package applications these days is using container technologies such as [Docker](https://www.docker.com).

> Using Docker's tooling, we can build and package the application as a Docker image, publish it in a Docker repository, and later launch it directly on a server, or in a platform that supports Docker deployments such as [Kuberenetes](https://kubernetes.io). Many public cloud provides including AWS, GCP, Azure, IBM and others encourage this kind of deployment.

Reviewer: ```suggestion
Using Docker's tooling, we can build and package the application as a Docker image, publish it in a Docker repository, and later launch it directly on a server, or in a platform that supports Docker deployments such as [Kubernetes](https://kubernetes.io). Many public cloud providers including AWS, GCP, Azure, IBM and others encourage this kind of deployment.
```

> Here is an example `Dockerfile` that builds and packages the application on top of a CentoOS 8 base images:

Reviewer: ```suggestion
Here is an example `Dockerfile` that builds and packages the application on top of a CentoOS 8 base image:
```

# Packaging Applications for Deployment

Once an application is built for production, it still needs to be packaged before it can be deployed to servers. There are several strategies for packaging Swift applications for deployment.

> 

Author: good point @ktoso, once we have enough content I think some of these will go into swift.org.

> 

Reviewer 2: Excellent! Let's stick to simple .md until then, then 👍

## Docker

One of the most popular ways to package applications these days is using container technologies such as [Docker](https://www.docker.com).

Using Docker's tooling, we can build and package the application as a Docker image, publish it to a Docker repository, and later launch it directly on a server or on a platform that supports Docker deployments such as [Kubernetes](https://kubernetes.io). Many public cloud providers including AWS, GCP, Azure, IBM and others encourage this kind of deployment.

> Here is an example `Dockerfile` that builds and packages the application on top of CentoOS:

Reviewer 3: ```suggestion
Here is an example `Dockerfile` that builds and packages the application on top of CentOS:
```

```Dockerfile
#------- build -------
FROM swiftlang/swift:nightly-centos8 as builder

# set up the workspace
RUN mkdir /workspace
WORKDIR /workspace

# copy the source to the docker image
COPY . /workspace

> 

Reviewer 3: @tomerd 99.9% sure that since 5.2 we don’t need the `-Xswiftc -g` anymore

Author: @aciidb0mb3r?

Reviewer 3: @tomerd yes, I'm correct. See this:

5.1 without `-g` leads to no source info:

```
$ jw-docker-swift-5.1 bash -c "rm -rf .build && swift build -c release  && lldb -o 'break set -n fcntl' -o run -o bt .build/release/tpkg"
[2/2] Linking tpkg
(lldb) target create ".build/release/tpkg"
Current executable set to '.build/release/tpkg' (x86_64).
(lldb) break set -n fcntl
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
3 locations added to breakpoint 1
1 location added to breakpoint 1
1 location added to breakpoint 1
Process 56 stopped
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
    frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32:
->  0x7ffff7fedd00 <+0>: pushq  %rbp
    0x7ffff7fedd01 <+1>: movq   %rsp, %rbp
    0x7ffff7fedd04 <+4>: xorl   %eax, %eax
    0x7ffff7fedd06 <+6>: popq   %rbp
Target 0: (tpkg) stopped.

Process 56 launched: '/tmp/tpkg/.build/release/tpkg' (x86_64)
(lldb) bt
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
    frame #1: 0x0000555555554a19 tpkg`main + 9
    frame #2: 0x00007ffff655cb97 libc.so.6`__libc_start_main + 231
    frame #3: 0x000055555555479a tpkg`_start + 42
```

but with `-g`

```
$ jw-docker-swift-5.1 bash -c "rm -rf .build && swift build -c release -Xswiftc -g && lldb -o 'break set -n fcntl' -o run -o bt .build/release/tpkg"
[2/2] Linking tpkg
(lldb) target create ".build/release/tpkg"
Current executable set to '.build/release/tpkg' (x86_64).
(lldb) break set -n fcntl
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
3 locations added to breakpoint 1
1 location added to breakpoint 1
1 location added to breakpoint 1
Process 56 stopped
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
    frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32:
->  0x7ffff7fedd00 <+0>: pushq  %rbp
    0x7ffff7fedd01 <+1>: movq   %rsp, %rbp
    0x7ffff7fedd04 <+4>: xorl   %eax, %eax
    0x7ffff7fedd06 <+6>: popq   %rbp
Target 0: (tpkg) stopped.

Process 56 launched: '/tmp/tpkg/.build/release/tpkg' (x86_64)
(lldb) bt
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
    frame #1: 0x0000555555554a19 tpkg`main at main.swift:8:1 [opt]
    frame #2: 0x00007ffff655cb97 libc.so.6`__libc_start_main + 231
    frame #3: 0x000055555555479a tpkg`_start + 42
```

In 5.2 however, even without `-g`:

```
$ jw-docker-swift-5.2 bash -c "rm -rf .build && swift build -c release  && lldb -o 'break set -n fcntl' -o run -o bt .build/release/tpkg"
[2/2] Linking tpkg
(lldb) target create ".build/release/tpkg"
Current executable set to '/tmp/tpkg/.build/release/tpkg' (x86_64).
(lldb) break set -n fcntl
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
3 locations added to breakpoint 1
1 location added to breakpoint 1
1 location added to breakpoint 1
Process 56 stopped
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
    frame #0: 0x00007ffff7fe9a00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32:
->  0x7ffff7fe9a00 <+0>: pushq  %rbp
    0x7ffff7fe9a01 <+1>: movq   %rsp, %rbp
    0x7ffff7fe9a04 <+4>: xorl   %eax, %eax
    0x7ffff7fe9a06 <+6>: popq   %rbp
Target 0: (tpkg) stopped.

Process 56 launched: '/tmp/tpkg/.build/release/tpkg' (x86_64)
(lldb) bt
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff7fe9a00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
    frame #1: 0x0000555555554a89 tpkg`main at main.swift:8:1 [opt]
    frame #2: 0x00007ffff6affb97 libc.so.6`__libc_start_main + 231
    frame #3: 0x00005555555547ba tpkg`_start + 42
(lldb) 
```

Nice! Source info and (what you can't see here): even colours in the terminal!

Author: @weissi cool, I will remove from there and we should update our readme.md as well with this info.

Reviewer 3: @tomerd done, https://github.com/swift-server/guides/pull/13

RUN swift build -c release -Xswiftc -g

#------- package -------
FROM centos:8
# copy executables
COPY --from=builder /workspace/.build/release/<executable-name> /
# copy Swift's dynamic libraries dependencies
COPY --from=builder /usr/lib/swift/linux/lib*so* /

# set the entry point (application name)
CMD ["<executable-name>"]
```

> To create the local Docker image from the `Dockerfile` use the `docker build` command from the application's source location, e.g.:

Reviewer: ```suggestion
To create a local Docker image from the `Dockerfile` use the `docker build` command from the application's source location, e.g.:
```

```bash
$ docker build . -t <my-app>:<my-app-version>
```

> To test the local image use the `docker run` command, .e.g.:

Reviewer: ```suggestion
To test the local image use the `docker run` command, e.g.:
```

```bash

> $ docker run <my-app>:<my-app-version>

Reviewer 2: May I suggest that we amend a bit the style we do code snippets with user filled in things a bit?

It's nice if before we start there's a

```
export MY_APP=<my-app> 
export MY_APP_VERSION=<...>
```

and then in all snippets we use those; this allows people to set their values and copy paste the steps if they want to.

(as silly as it is to encourage copy paste i think it helps some folks and eases the barrier to try things out).

Just an idea

```

> Finally, use the `docker push` command to publish the application's Docker image to a Docker repository of your choice. e.g.:

Reviewer 2: ```suggestion
Finally, use the `docker push` command to publish the application's Docker image to a Docker repository of your choice, e.g.:
```

Closer to proper usage of it at end of sentence 🤔 
https://english.stackexchange.com/questions/42326/the-use-of-e-g-at-the-end-of-a-sentence

> Finally, use the `docker push` command to publish the Application's Docker image to a Docker repository of your choice. e.g.:

Reviewer: ```suggestion
Finally, use the `docker push` command to publish the application's Docker image to a Docker repository of your choice. e.g.:
```

```bash
$ docker tag <my-app>:<my-app-version> <docker-hub-user>/<my-app>:<my-app-version>
$ docker push <docker-hub-user>/<my-app>:<my-app-version>
```

> At this point, the application's Docker image is ready to be deployed to the server hosts (which need to run docker), or to one of the platforms that support Docker deployments.

Reviewer: ```suggestion
At this point, the application's Docker image is ready to be deployed to the server hosts (which need to run docker), or to one of the platforms that supports Docker deployments.
```

Since distroless supports Docker and is based on Debian, packaging a Swift application on it is fairly similar to the Docker process above. Here is an example `Dockerfile` that builds and packages the application on top of a distroless's C++ base image:

```Dockerfile
#------- build -------
# Building using Ubuntu Bionic since its compatible with Debian runtime
FROM swiftlang/swift:nightly-bionic as builder

# set up the workspace
RUN mkdir /workspace
WORKDIR /workspace

# copy the source to the docker image
COPY . /workspace

RUN swift build -c release -Xswiftc -g

#------- package -------
# Running on distroless C++ since it includes
# all(*) the runtime dependencies Swift programs need
FROM gcr.io/distroless/cc-debian10
# copy executables
COPY --from=builder /workspace/.build/release/<executable-name> /
# copy Swift's dynamic libraries dependencies
COPY --from=builder /usr/lib/swift/linux/lib*so* /

# set the entry point (application name)
CMD ["<executable-name>"]
```

> Note the above this uses `gcr.io/distroless/cc-debian10` as the runtime image which should work for Swift programs that do not use `FoundationNetworking` or `FoundationXML`. In order to provide a more complete support we (the community) could put in a PR into distroless to introduce a base image for Swift that includes `libcurl` and `libxml` which are required for `FoundationNetworking` and `FoundationXML` respectively.

Reviewer: ```suggestion
Note the above uses `gcr.io/distroless/cc-debian10` as the runtime image which should work for Swift programs that do not use `FoundationNetworking` or `FoundationXML`. In order to provide more complete support we (the community) could put in a PR into distroless to introduce a base image for Swift that includes `libcurl` and `libxml` which are required for `FoundationNetworking` and `FoundationXML` respectively.
```

> ## Archive (Tarball, Zipfile, etc)

Reviewer: ```suggestion
## Archive (Tarball, ZIP file, etc.)
```

> Since cross-compiling Swift for Linux is not [yet] supported on Mac or Windows, we need to use virtualization technologies like Docker to compile applications we are targeting to run on Linux.

Reviewer 2: ```suggestion
Since cross-compiling Swift for Linux is not (yet) supported on Mac or Windows, we need to use virtualization technologies like Docker to compile applications we are targeting to run on Linux.
```

?

> Since cross compiling Swift for Linux is not [yet] supported on Mac or Windows, we need to use use virtualization technologies like Docker to compile applications we are targeting to run on Linux.

Reviewer: ```suggestion
Since cross-compiling Swift for Linux is not [yet] supported on Mac or Windows, we need to use virtualization technologies like Docker to compile applications we are targeting to run on Linux.
```

> That said, this does not mean we must also package the applications as Docker images in order to deploy them. While using Docker images for deployment is convenient and popular, an application can also be packaged using a simple and lightweight archive format like tarball or zipfile, then uploaded to the server where it can be extracted and run.

Reviewer: ```suggestion
That said, this does not mean we must also package the applications as Docker images in order to deploy them. While using Docker images for deployment is convenient and popular, an application can also be packaged using a simple and lightweight archive format like tarball or ZIP file, then uploaded to the server where it can be extracted and run.
```

Here is an example of using Docker and `tar` to build and package the application for deployment on Ubuntu servers:

> First, use the `docker run` command from the application's source location to build the application:

Reviewer: ```suggestion
First, use the `docker run` command from the application's source location to build it:
```

```bash
$ docker run --rm \
  -v "$PWD:/workspace" \
  -w /workspace \
  swift:5.2-bionic \

>   /bin/bash -cl "swift build -c release"

Reviewer 3: ```suggestion
  /bin/bash -xcl "swift -version && swift build -c release"
```

feel free to ignore but I prefer to always see the version number as that's often important.

Reviewer 3: I also like `-x` because it prints all the random env variables that may be important when debugging stuff (but it does make the output a little bit uglier ofc)

Reviewer 3: sample output:

```
$ jw-docker-swift-5.2 /bin/bash -xcl "swift -version && swift build -c release"
+ '[' '' ']'
+ '[' -d /etc/profile.d ']'
+ for i in /etc/profile.d/*.sh
+ '[' -r /etc/profile.d/01-locale-fix.sh ']'
+ . /etc/profile.d/01-locale-fix.sh
+++ /usr/bin/locale-check C.UTF-8
++ eval
+ unset i
+ '[' /bin/bash ']'
+ '[' -f /root/.bashrc ']'
+ . /root/.bashrc
++ '[' -z '' ']'
++ return
+ mesg n
+ true
+ swift -version
Swift version 5.2.3 (swift-5.2.3-RELEASE)
Target: x86_64-unknown-linux-gnu
+ swift build -c release
[39/39] Linking NIOPerformanceTester
```

```

> Note we are bind mounting the source directory so that the build writes the  build artifacts to the local drive from which we will package them later.

Reviewer: ```suggestion
Note we are bind mounting the source directory so that the build writes the build artifacts to the local drive from which we will package them later.
```

Next we can create a staging area with the application's executables and Swift's dynamic libraries dependencies:

```bash
$ docker run --rm \
  -v `pwd`:/workspace \
  -w /workspace \
  swift:5.2-bionic  \
  /bin/bash -cl ' \
     rm -rf .build/install && mkdir -p .build/install; \

>      cp -P .build/release/<executable-name> .build/install/; \

Reviewer 3: the `;`s should be `&&`s, no?

Author: for failure case sure

     cp -P /usr/lib/swift/linux/lib*so* .build/install/'
```

> Note this command could be combined with the build command above, we separated them to make the example more readable.

Reviewer: ```suggestion
Note this command could be combined with the build command above--we separated them to make the example more readable.
```

Finally, create a tarball from the staging directory:

```bash
$ tar cvzf <my-app>-<my-app-version>.tar.gz -C .build/install .
```

We can test the integrity of the tarball by extracting it to a directory and running the application in a Docker runtime container:

```bash
$ cd <extracted directory>

> $ docker run -v `pwd`:/app -w /app swift:5.2-bionic-slim ./<executable-name>

Reviewer 3: - I think we should quote this just in case the dir has a space
- I think we shouldn't recommend the outdated ``` ` ```

So what about

    docker run -v "$(pwd)":/app -w /app swift:5.2-bionic-slim ./<executable-name>

Personally, I actually use

    docker run -v "$PWD:/app" -w /app swift:5.2-bionic-slim ./<executable-name>

Author: sure. my age shows :)

```

> Deploying the application's tarball to the target server can be done using utilities like `scp`, or in a more sophisticated setup using configuration management system like `chef`, `puppet`, `ansible`,  etc

Reviewer: ```suggestion
Deploying the application's tarball to the target server can be done using utilities like `scp`, or in a more sophisticated setup using configuration management system like `chef`, `puppet`, `ansible`, etc.
```

## Source Distribution

Another distribution technique popular with dynamic languages like Ruby or Javascript is distributing the source to the server, then compiling it on the server itself.

To build Swift applications directly on the server, the server must have the correct Swift toolchain installed. [Swift.org](https://swift.org/download/#linux) publishes toolchains for a variety of Linux distributions, make sure to use the one matching your server Linux version and desired Swift version.

The main advantage of this approach is that it is easy. Additional advantage is the server has the full toolchain (e.g. debugger) that can help troubleshoot issues "live" on the server.

> The main disadvantage of this approach that the server has the full toolchain (e.g. compiler) which means a sophisticated attacker can potentially find ways to execute code. They can also potentially gain access to the source code which might be sensitive. If the application code needs to cloned from a private or protected repository, the server needs access to credentials which adds additional attack surface area.

Reviewer: ```suggestion
The main disadvantage of this approach that the server has the full toolchain (e.g. compiler) which means a sophisticated attacker can potentially find ways to execute code. They can also potentially gain access to the source code which might be sensitive. If the application code needs to be cloned from a private or protected repository, the server needs access to credentials which adds additional attack surface area.
```
